### ISSUE TITLE
Decode `Transfer-Encoding: chunked` responses

### ISSUE THREAD

**Issue by anjackson** [ukwa/webarchive-discovery#218]
Running the MDX generator (which is like the Solr indexer but just outputs JSONL files) rather than pushing to Solr), I'm seeing this error:

```
java.io.IOException: Resetting to invalid mark
	at java.io.BufferedInputStream.reset(BufferedInputStream.java:448)
	at org.apache.commons.io.input.ProxyInputStream.reset(ProxyInputStream.java:167)
	at uk.bl.wa.nanite.droid.InputStreamByteReader.readByte(InputStreamByteReader.java:52)
	at net.domesdaybook.matcher.sequence.searcher.BoyerMooreHorspoolSearcher.searchBackwards(BoyerMooreHorspoolSearcher.java:140)
	at uk.gov.nationalarchives.droid.core.signature.droid6.SubSequence.findSequenceFromPosition(SubSequence.java:660)
	at uk.gov.nationalarchives.droid.core.signature.droid6.ByteSequence.matches(ByteSequence.java:413)
	at uk.gov.nationalarchives.droid.core.signature.droid6.InternalSignature.matches(InternalSignature.java:399)
	at uk.gov.nationalarchives.droid.core.signature.droid6.InternalSignatureCollection.getMatchingSignatures(InternalSignatureCollection.java:99)
	at uk.gov.nationalarchives.droid.core.signature.droid6.FFSignatureFile.runFileIdentification(FFSignatureFile.java:495)
	at uk.gov.nationalarchives.droid.core.BinarySignatureIdentifier.matchBinarySignatures(BinarySignatureIdentifier.java:103)
	at uk.bl.wa.nanite.droid.DroidDetector.getPUIDs(DroidDetector.java:449)
	at uk.bl.wa.nanite.droid.DroidDetector.detectPUIDs(DroidDetector.java:424)
	at uk.bl.wa.nanite.droid.DroidDetector.detect(DroidDetector.java:377)
	at uk.bl.wa.analyser.payload.DroidDetectorAnalyser.analyse(DroidDetectorAnalyser.java:113)
	at uk.bl.wa.analyser.WARCPayloadAnalysers.analyse(WARCPayloadAnalysers.java:109)
	at uk.bl.wa.indexer.WARCIndexer.extract(WARCIndexer.java:415)
	at uk.bl.wa.indexer.WARCIndexer.extract(WARCIndexer.java:273)
	at uk.bl.wa.hadoop.indexer.WARCIndexerMapper.innerMap(WARCIndexerMapper.java:160)
	at uk.bl.wa.hadoop.indexer.mdx.WARCMDXMapper.map(WARCMDXMapper.java:82)
	at uk.bl.wa.hadoop.indexer.mdx.WARCMDXMapper.map(WARCMDXMapper.java:50)
	at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:50)
	at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:391)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:325)
	at org.apache.hadoop.mapred.Child$4.run(Child.java:270)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1127)
	at org.apache.hadoop.mapred.Child.main(Child.java:264)
```

Which implies there's some issues managing buffering of InputStreams and marks. Worth checking if this crops up with PDFs only?

Also,

- Hash check errors need to be reported properly - I saw some (likely due to tmp file problems rather than data problems) but wasn't sure they were being reported.
- Had to set a PDFBox tmp/cache location.
- Should update Nanite.

**Comment by anjackson** [comment]
I'm also seeing some alleged hash-calculation failures. I think something is going wrong with the input streams. The `1_data/ethos/warcs/WARCPROX-20200404114954393-00474-mja43xl7.warc.gz` file seems to have an example near the top, which is useful.

```
2020-04-09 12:24:29 INFO  SolrRecordFactory:77 - Created SolrRecordFactory(defaultMaxLength=-1, maxLengths=[content=524288, url_norm=2048, links=2048, url=2048])
2020-04-09 12:24:29 INFO  WARCIndexerMapper:99 - Got task.id attempt_202002261158_0330_m_000010_0 and input.file hdfs://namenode:54310/1_data/ethos/warcs/WARCPROX-20200405051944929-01309-mja43xl7.warc.gz
2020-04-09 12:24:32 ERROR HashedCachedInputStream:162 - Hashes are not equal for this input!
2020-04-09 12:24:32 ERROR HashedCachedInputStream:163 -  - payload hash from header = sha1:TSEG4C6T5W555ZESUXHCWUZY2PVOXKBS
2020-04-09 12:24:32 ERROR HashedCachedInputStream:165 -  - payload hash from content = sha1:NC3IOQZ4OG2KJGNQR5477T37JXUJCINJ
2020-04-09 12:24:32 ERROR HashedCachedInputStream:184 - Hashing: https://purehost.bath.ac.uk/ws/portalfiles/portal/187906421/M_Langenhan_Final_DS_Printed_PhD_Thesis_2017.pdf@1492
java.lang.RuntimeException: Hash check failed!
	at uk.bl.wa.util.HashedCachedInputStream.init(HashedCachedInputStream.java:166)
	at uk.bl.wa.util.HashedCachedInputStream.<init>(HashedCachedInputStream.java:106)
	at uk.bl.wa.indexer.WARCIndexer.extract(WARCIndexer.java:381)
	at uk.bl.wa.indexer.WARCIndexer.extract(WARCIndexer.java:273)
	at uk.bl.wa.hadoop.indexer.WARCIndexerMapper.innerMap(WARCIndexerMapper.java:162)
	at uk.bl.wa.hadoop.indexer.mdx.WARCMDXMapper.map(WARCMDXMapper.java:82)
	at uk.bl.wa.hadoop.indexer.mdx.WARCMDXMapper.map(WARCMDXMapper.java:50)
	at org.apache.hadoop.mapred.MapRunner.run(MapRunner.java:50)
	at org.apache.hadoop.mapred.MapTask.runOldMapper(MapTask.java:391)
	at org.apache.hadoop.mapred.MapTask.run(MapTask.java:325)
	at org.apache.hadoop.mapred.Child$4.run(Child.java:270)
	at java.security.AccessController.doPrivileged(Native Method)
	at javax.security.auth.Subject.doAs(Subject.java:422)
	at org.apache.hadoop.security.UserGroupInformation.doAs(UserGroupInformation.java:1127)
	at org.apache.hadoop.mapred.Child.main(Child.java:264)
```

**Comment by anjackson** [comment]
Having poked around with `warcio`, it's clear these are `Transfer-Encoding: chunked` records, and that's causing havoc.  Neither `warcio` nor this project de-chunks the payload, so the parsers go wrong and the validation fails.

Here's an example:

```
WARC/1.0^M
WARC-Type: response^M
WARC-Record-ID: <urn:uuid:09959aa4-b44b-4b93-903a-18964f062e9b>^M
WARC-Date: 2020-04-04T11:46:32Z^M
WARC-Target-URI: https://www.escholar.manchester.ac.uk/api/datastream?publicationPid=uk-ac-man-scw:212377&datastreamId=FULL-TEXT.PDF^M
WARC-IP-Address: 130.88.212.119^M
Content-Type: application/http;msgtype=response^M
WARC-Payload-Digest: sha1:FNXWBUS7VMY4GF4WOZKRYFDT4FUWIZD4^M
Content-Length: 2649217^M
WARC-Block-Digest: sha1:VXY7X743WSNYZCZH4WIM6B6MQGZKMNE3^M
^M
HTTP/1.1 200 OK^M
Date: Sat, 04 Apr 2020 11:46:32 GMT^M
Server: Apache^M
Content-Type: application/pdf;charset=UTF-8^M
X-Frame-Options: SAMEORIGIN^M
Transfer-Encoding: chunked^M
^M
1ff8^M
%PDF-1.5^M%âãÏÓ^M
```

The `WARC-Block-Digest` could be checked without de-chunking the payload, but we need to de-chunk it anyway so we can parse the darn thing.

**Comment by anjackson** [comment]
Note that this issue has already been raised on `warcio` here: https://github.com/webrecorder/warcio/issues/75

**Comment by anjackson** [comment]
For this implementation, we should be able to wrap the original stream in a [ChunkedInputStream](https://hc.apache.org/httpclient-3.x/apidocs/org/apache/commons/httpclient/ChunkedInputStream.html) and then cache that.

**Comment by anjackson** [comment]
Well, the decoding works, but:

- It's not certain it's exactly reconstructing the original, as the hashes do not match for the files used in the integration tests. However, it's not 100% clear if this is a problem with those (older) files, rather than the decoding. Running again over modern WARCs should help rule this out.
- DROID is still complaining, but possibly that's a separate issue? Or possibly to do with the content length being unknown in the current implementation?

**Comment by tokee** [comment]
I am currently investigating this issue from the perspective of webrecorder-WARCs. As far as I can see, webrecorder leaves the `Transfer-Encoding: chunked` header but de-chunks the body. This means that your fix 410fba90d483ef042b62ac0e4a2c2eaeb18928ef breaks chunked webrecorder-WARCs.

As far as I can see, there are no clear way to see if the body has been de-chunked or not. So what to do? Maybe make the `ChunkedInputStream` look for `^[0-9a-z]{1,8}(;.*)?\r\n` at the beginning of the stream and just pass the stream (after resetting to mark) if it does not match?

We really should have unit-tests for this stuff. There are far too many dragons here.

**Comment by tokee** [comment]
Should this be closed? It was addressed in #220, but I don't know if it fixes the problems @anjackson encountered?
